integration-tests/src/test/java/com/datastax/oss/driver/core/cql/PreparedStatementCachingIT.java

```diff
@@ -0,0 +1,401 @@
+          .build();
+
+  @Rule public TestRule chain = RuleChain.outerRule(ccmRule).around(sessionRule);
+
+  private static class PreparedStatementRemovalEvent {
+
+    private final ByteBuffer queryId;
+
+    public PreparedStatementRemovalEvent(ByteBuffer queryId) {
+      this.queryId = queryId;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || !(o instanceof PreparedStatementRemovalEvent)) return false;
+      PreparedStatementRemovalEvent that = (PreparedStatementRemovalEvent) o;
+      return Objects.equals(queryId, that.queryId);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(queryId);
+    }
+
+    @Override
+    public String toString() {
+      return "PreparedStatementRemovalEvent{" + "queryId=" + queryId + '}';
+    }
+  }
+
+  private static class TestCqlPrepareAsyncProcessor extends CqlPrepareAsyncProcessor {
+
+    private static final Logger LOG =
+        LoggerFactory.getLogger(PreparedStatementCachingIT.TestCqlPrepareAsyncProcessor.class);
+
+    private static RemovalListener<PrepareRequest, CompletableFuture<PreparedStatement>>
+        buildCacheRemoveCallback(@NonNull Optional<DefaultDriverContext> context) {
+      return (evt) -> {
+        try {
+          CompletableFuture<PreparedStatement> future = evt.getValue();
+          ByteBuffer queryId = Uninterruptibles.getUninterruptibly(future).getId();
+          context.ifPresent(
+              ctx -> ctx.getEventBus().fire(new PreparedStatementRemovalEvent(queryId)));
+        } catch (Exception e) {
+          LOG.error("Unable to register removal handler", e);
+        }
+      };
+    }
+
+    public TestCqlPrepareAsyncProcessor(@NonNull Optional<DefaultDriverContext> context) {
+      // Default CqlPrepareAsyncProcessor uses weak values here as well.  We avoid doing so
+      // to prevent cache entries from unexpectedly disappearing mid-test.
+      super(
+          CacheBuilder.newBuilder().removalListener(buildCacheRemoveCallback(context)).build(),
+          context);
+    }
+  }
+
+  private static class TestDefaultDriverContext extends DefaultDriverContext {
+    public TestDefaultDriverContext(
+        DriverConfigLoader configLoader, ProgrammaticArguments programmaticArguments) {
+      super(configLoader, programmaticArguments);
+    }
+
+    @Override
+    protected RequestProcessorRegistry buildRequestProcessorRegistry() {
+      List<RequestProcessor<?, ?>> processors =
+          BuiltInRequestProcessors.createDefaultProcessors(this);
+      processors.removeIf((processor) -> processor instanceof CqlPrepareAsyncProcessor);
+      processors.removeIf((processor) -> processor instanceof CqlPrepareSyncProcessor);
+      CqlPrepareAsyncProcessor asyncProcessor = new TestCqlPrepareAsyncProcessor(Optional.of(this));
+      processors.add(2, asyncProcessor);
+      processors.add(3, new CqlPrepareSyncProcessor(asyncProcessor));
```

Comment:
A question: do we need some comments here explaining why we're doing these manipulations of the RequestProcessorRegistry?  Perhaps it's obvious but it seemed at least worth asking about...